aptos-move/framework/src/natives/state_storage.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_state_view::state_storage_usage::StateStorageUsage;
+use aptos_types::vm_status::StatusCode;
+use better_any::{Tid, TidAble};
+use move_deps::move_binary_format::errors::PartialVMError;
+use move_deps::move_vm_types::{
+    loaded_data::runtime_types::Type,
+    natives::function::NativeResult,
+    values::{Struct, Value},
+};
+use move_deps::{
+    move_binary_format::errors::PartialVMResult,
+    move_vm_runtime::native_functions::{NativeContext, NativeFunction},
+};
+use smallvec::smallvec;
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+/// Ability to reveal the state storage utilization info.
+pub trait StateStorageUsageResolver {
+    fn get_state_storage_usage(&self) -> anyhow::Result<StateStorageUsage>;
+}
+
+/// Exposes the ability to query state storage utilization info to native functions.
+#[derive(Tid)]
+pub struct NativeStateStorageContext<'a> {
+    resolver: &'a dyn StateStorageUsageResolver,
+}
+
+impl<'a> NativeStateStorageContext<'a> {
+    pub fn new(resolver: &'a dyn StateStorageUsageResolver) -> Self {
+        Self { resolver }
+    }
+}
+
+/***************************************************************************************************
+ * native get_state_storage_usage_only_at_eopch_beginning
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct GetUsageGasParameters {
+    pub base_cost: u64,
```

Comment:
@msmouse you need to define an on-chain gas entry for this in the `aptos-gas` crate. See `aptos_framework.rs` and `move_stdlib.rs` for examples.